### PR TITLE
Initialization with Job API

### PR DESCRIPTION
**Pull request by richtja**
Right now when we want to run tests through avocado Job API the avocado-vt is
not loaded, because a lot of avocado-vt settings is initialized through CLI
plugin.  This commit moves the initialization to the Init plugin and leaves CLI
plugin just for necessary command line setup

Signed-off-by: Jan Richter <jarichte@redhat.com>

### CODE REVIEW

**Review by willianrampazzo — APPROVED** [review]

File: avocado_vt/loader.py [hunk 1]
```diff
@@ -134,12 +134,12 @@ def _save_parser_cartesian_config(self, parser):
                 cartesian_config.write("%s\n" % statement)
 
     def get_extra_listing(self):
-        if get_opt(self.config, 'vt_list_guests'):
```
Comment (clebergnu) [inline avocado_vt/loader.py:137]: Nice catch.  I think these two fixes deserve their own commit.

File: avocado_vt/plugins/vt.py [hunk 2]
```diff
@@ -18,20 +18,19 @@
...
 from avocado.utils import path as utils_path
 
-from virttest import data_dir
-from virttest import defaults
-from virttest import standalone_test
+from virttest import data_dir, defaults, standalone_test
 from virttest.compat import get_settings_value
-from virttest.standalone_test import SUPPORTED_TEST_TYPES
-from virttest.standalone_test import SUPPORTED_LIBVIRT_URIS
+from virttest.standalone_test import (SUPPORTED_LIBVIRT_URIS,
```
Comment (clebergnu) [inline avocado_vt/plugins/vt.py:29]: It's *better* to keep such style changes (an improvement in this case) in a separate commit.

File: avocado_vt/plugins/vt.py [hunk 3]
```diff
@@ -42,59 +41,109 @@
                       "plugin to get rid of this message")
 
 
+def add_option(parser, namespace, arg, default=None, action='store',
+               help_msg='', nargs=None):
+    if hasattr(Settings, 'add_argparser_to_option'):
```
Comment (clebergnu) [inline avocado_vt/plugins/vt.py:46]: `virttest.compat` seems an appropriate place for this utility, because there's already a utility for checking if the settings is "new or old" style (`is_registering_settings_required()`), and also a conditional block in which you can add two versions of this `add_option()` utility.  There's already `get_opt()`, `set_opt()`, `set_opt_from_settings()` and `get_settings_value()` there.

File: avocado_vt/plugins/vt_init.py [hunk 4]
```diff
@@ -177,8 +234,26 @@ def initialize(self):
...
+                                     nargs='+', default=None,
+                                     help_msg=help_msg)
+
             # [plugins.vtjoblock] section
             help_msg = 'Directory in which to write the lock file'
             settings.register_option('plugins.vtjoblock', 'dir',
                                      help_msg=help_msg,
                                      default='/tmp')
+
+            virt_loader = getattr(importlib.import_module('avocado_vt.loader'),
```
Comment (clebergnu) [inline avocado_vt/plugins/vt_init.py:257]: I'm not following the reason to use `importlib` here.  Why not simply:

```python
from ..loader import VirtTestLoader
from avocado.core import loader

loader.register_plugin(VirtTestLoader)
```

File: avocado_vt/plugins/vt_list.py [hunk 5]
```diff
@@ -96,22 +98,44 @@ def configure(self, parser):
...
+                                                help=help_list_guests)
+            vt_compat_group_lister.add_argument("--vt-list-archs", default=False,
+                                                action="store_true",
+                                                help=help_list_archs)
         add_basic_vt_options(vt_compat_group_lister)
         add_qemu_bin_vt_option(vt_compat_group_lister)
 
     def run(self, config):
-        loader.register_plugin(VirtTestLoader)
+        if not hasattr(plugin_interfaces, 'Init'):
```
Comment (clebergnu) [inline avocado_vt/plugins/vt_list.py:140]: It won't hurt to *attempt* to register the loader plugin multiple times, so you can drop the conditional check.

File: avocado_vt/plugins/vt_list.py [hunk 6]
```diff
@@ -96,22 +98,44 @@ def configure(self, parser):
...
-                                            "--vt-config --vt-machine-type "
-                                            "and --vt-arch args are ignored)")
+
+        help_list_guests = ("Also list the available guests (this option "
+                            "ignores the --vt-config and --vt-guest-os)")
+        help_list_archs = ("Also list the available arch/machines for the "
+                           "given guest OS. (Use \"--vt-guest-os ''\" to see "
+                           "all combinations; --vt-config --vt-machine-type "
+                           "and --vt-arch args are ignored)")
+        if hasattr(Settings, 'register_option'):
```
Comment (clebergnu) [inline avocado_vt/plugins/vt_list.py:108]: This seems like another good candidate for a `virttest.compat` utility function.

**Review by clebergnu — CHANGES_REQUESTED** [review]
Hi @richtja, thanks for this! It actually contains some similarities to what I had to do while working on the VT resolver, but this work of yours cover a lot more!

I make a number of comments/suggestions/requests here.

File: avocado_vt/plugins/vt.py (hunk 2 above)
Comment (richtja) [reply avocado_vt/plugins/vt.py:29]: Sure I will split the commit

File: avocado_vt/plugins/vt.py (hunk 3 above)
Comment (richtja) [reply avocado_vt/plugins/vt.py:46]: Cool, I didn't know about this. I will use it

File: avocado_vt/plugins/vt_init.py (hunk 4 above)
Comment (richtja) [reply avocado_vt/plugins/vt_init.py:257]: Unfortunately I can't do global import here. Because the `VirtTestLoader` have to be registered after initialization because of dependencies. And I think that the local import is better via `importlib`.

File: avocado_vt/plugins/vt_list.py (hunk 5 above)
Comment (richtja) [reply avocado_vt/plugins/vt_list.py:140]: Yes, that is true. I will change that.

**Comment by richtja** [comment]
@clebergnu Thanks for your comments I will use them and create a new version.

**Comment by richtja** [comment]
Closing in favor of #2844
